### PR TITLE
[wmco] Add validation for fields in the CRD

### PR DESCRIPTION
**Pull request by akhil-rane**
This PR updates the WMC CRD to set the minimum value of 'replicas' 
field to 0 by executing 'operator-sdk generate crds' command

It also adds validations for 'replicas' field in the CRD. Following are the
validations that are performed.

- 'replicas' field is a required field and should be present
- 'replicas' field cannot have value less than 0

### CODE REVIEW

File: test/e2e/wmco_test.go [hunk 1]
```diff
@@ -25,6 +26,17 @@ var (
...
 
+const (
+	// api version of WindowsMachineConfig custom resource
+	wmcApiVersion = "wmc.openshift.io/v1alpha1"
+	// account ID of AWS account
+	awsAccountID = "default"
+	// sshKey is the key that will be used to access created Windows VMs
+	sshKey = "libra"
+	// ec2 instance type of the Windows VMs
+	ec2InstanceType = "m4.large"
```
Comment (sebsoto) [inline test/e2e/wmco_test.go]: docs should start with he variable name

File: test/e2e/wmco_test.go [hunk 2]
```diff
@@ -33,6 +45,7 @@ func TestWMCO(t *testing.T) {
 	// TODO: In future, we'd like to skip the teardown for each test. As of now, since we just have deletion it should
 	// 		be ok to call destroy directly.
 	//		Jira Story: https://issues.redhat.com/browse/WINC-283
+	t.Run("validations of WindowsMachineConfigs CRD", testWMCOCustomResourceDefinitionValidations)
```
Comment (sebsoto) [inline test/e2e/wmco_test.go]: validations -> validation

File: test/e2e/wmco_test.go (hunk 1 above)
Comment (sebsoto) [inline test/e2e/wmco_test.go]: m5a.large

File: test/e2e/wmco_test.go [hunk 3]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
 	}
 	defer testCtx.Cleanup()
 }
+
+// testWMCOCustomResourceDefinitionValidations tests if validations of the fields of WindowsMachineConfigs CRD are
+// working as expected
+func testWMCOCustomResourceDefinitionValidations(t *testing.T) {
+	testCtx := framework.NewTestCtx(t)
+	namespace, err := testCtx.GetNamespace()
+	if err != nil {
```
Comment (sebsoto) [inline test/e2e/wmco_test.go]: use require.NotNil

File: test/e2e/wmco_test.go [hunk 4]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+	}
+
+	err = framework.Global.Client.Create(context.TODO(), wmcoWithInvalidReplicasField,
+		&framework.CleanupOptions{TestContext: testCtx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	// Failing the test if WMCO custom resource is not throwing an error when value of replicas field is less than 0
+	require.NotEmpty(t, err, "Creation of WMCO custom resource did not throw an error when value of the "+
+		"replicas field is invalid")
+	// Failing the test if WMCO custom resource is throwing an error, but the error is not related to replicas field
+	// value being less than 0
+	require.Contains(t, err.Error(), "spec.replicas in body should be greater than or equal to 0",
```
Comment (sebsoto) [inline test/e2e/wmco_test.go]: This should be assert as its what were testing for and not required for something else

File: test/e2e/wmco_test.go [hunk 5]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+	}
+
+	err = framework.Global.Client.Create(context.TODO(), wmcoWithoutReplicasField,
+		&framework.CleanupOptions{TestContext: testCtx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	// Failing the test if WMCO custom resource is not throwing an error when replicas field is absent
+	require.NotEmpty(t, err, "Creation of WMCO custom resource did not throw an error when required replicas "+
+		"field is absent")
+	// Failing the test if WMCO custom resource is throwing an error, but the error is not related to replicas field
+	// being absent
+	require.Contains(t, err.Error(), "spec.replicas: Required value",
```
Comment (sebsoto) [inline test/e2e/wmco_test.go]: This should be assert as its what were testing for and not required for something else

File: test/e2e/wmco_test.go [hunk 6]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+		Spec: operator.WindowsMachineConfigSpec{
+			InstanceType: ec2InstanceType,
+			AWS:          &operator.AWS{CredentialAccountID: awsAccountID, SSHKeyPair: sshKey},
+		},
+	}
+
+	err = framework.Global.Client.Create(context.TODO(), wmcoWithoutReplicasField,
+		&framework.CleanupOptions{TestContext: testCtx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	// Failing the test if WMCO custom resource is not throwing an error when replicas field is absent
+	require.NotEmpty(t, err, "Creation of WMCO custom resource did not throw an error when required replicas "+
```
Comment (aravindhp) [inline test/e2e/wmco_test.go]: `require.Error()`

File: test/e2e/wmco_test.go [hunk 7]
```diff
@@ -25,6 +26,17 @@ var (
 	cleanupTimeout       = time.Second * 5
 )
 
+const (
+	// api version of WindowsMachineConfig custom resource
+	wmcApiVersion = "wmc.openshift.io/v1alpha1"
+	// account ID of AWS account
+	awsAccountID = "default"
+	// sshKey is the key that will be used to access created Windows VMs
+	sshKey = "libra"
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: unnecessary changes in this PR

File: test/e2e/wmco_test.go [hunk 8]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+// working as expected
+func testWMCOCustomResourceDefinitionValidations(t *testing.T) {
+	testCtx := framework.NewTestCtx(t)
+	namespace, err := testCtx.GetNamespace()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// create WMCO custom resource without replicas field
+	wmcoWithoutReplicasField := &operator.WindowsMachineConfig{
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: Are these just -ve scenarios where as the +ve scenario is somewhere else? Can you make this table driven?

File: test/e2e/wmco_test.go [hunk 9]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+	// Failing the test if WMCO custom resource is not throwing an error when replicas field is absent
+	require.NotEmpty(t, err, "Creation of WMCO custom resource did not throw an error when required replicas "+
+		"field is absent")
+	// Failing the test if WMCO custom resource is throwing an error, but the error is not related to replicas field
+	// being absent
+	require.Contains(t, err.Error(), "spec.replicas: Required value",
+		"Creation of WMCO custom resource threw an error that is not related to replicas field is being absent")
+
+	// create WMCO custom resource with replicas Field value set to -1
+	wmcoWithInvalidReplicasField := &operator.WindowsMachineConfig{
```
Comment (aravindhp) [inline test/e2e/wmco_test.go]: `wmc` and not `wmco`

File: test/e2e/wmco_test.go [hunk 10]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+
+	err = framework.Global.Client.Create(context.TODO(), wmcoWithoutReplicasField,
+		&framework.CleanupOptions{TestContext: testCtx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	// Failing the test if WMCO custom resource is not throwing an error when replicas field is absent
+	require.NotEmpty(t, err, "Creation of WMCO custom resource did not throw an error when required replicas "+
+		"field is absent")
+	// Failing the test if WMCO custom resource is throwing an error, but the error is not related to replicas field
+	// being absent
+	require.Contains(t, err.Error(), "spec.replicas: Required value",
+		"Creation of WMCO custom resource threw an error that is not related to replicas field is being absent")
```
Comment (aravindhp) [inline test/e2e/wmco_test.go]: `WMC CR` and not `WMCO CR`.

WMC == Windows Machine Config (API)
WMCO == Windows Machine Config Operator (Handler for the API)

File: test/e2e/wmco_test.go [hunk 11]
```diff
@@ -131,3 +144,67 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+			InstanceType: ec2InstanceType,
+			AWS:          &operator.AWS{CredentialAccountID: awsAccountID, SSHKeyPair: sshKey},
+			Replicas:     -1,
+		},
+	}
+
+	err = framework.Global.Client.Create(context.TODO(), wmcoWithInvalidReplicasField,
+		&framework.CleanupOptions{TestContext: testCtx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+	// Failing the test if WMCO custom resource is not throwing an error when value of replicas field is less than 0
+	require.NotEmpty(t, err, "Creation of WMCO custom resource did not throw an error when value of the "+
```
Comment (aravindhp) [inline test/e2e/wmco_test.go]: `require.Error()`

File: test/e2e/wmco_test.go (hunk 7 above)
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: They were getting repeated at too many places so I thought of making them constants

**Comment by akhil-rane** [comment]
> Thanks for working on this, @akhil-rane.
> 
> I just have some minor comments. One question, I've is `operator-sdk generate crds` overriding the changes you made considering you're modifying the CRD directly?



> Thanks for working on this, @akhil-rane.
> 
> I just have some minor comments. One question, I've is `operator-sdk generate crds` overriding the changes you made considering you're modifying the CRD directly?

yes, It is overriding my changes

File: test/e2e/wmco_test.go (hunk 7 above)
Comment (aravindhp) [reply test/e2e/wmco_test.go]: I am a +1 on this given he would have to copy paste them otherwise. Alternatively making the constants could have been a separate commit.

**Comment by aravindhp** [comment]
>    I just have some minor comments. One question, I've is operator-sdk generate crds overriding the changes you made considering you're modifying the CRD directly?
> yes, It is overriding my changes

From what I understand, you cannot keep running `operator-sdk generate crds` after you start making invasive changes to the CRD. But best to get confirmation from the SDK team.

**Comment by akhil-rane** [comment]
> Thanks for working on this, @akhil-rane.
> 
> I just have some minor comments. One question, I've is `operator-sdk generate crds` overriding the changes you made considering you're modifying the CRD directly?

@ravisantoshgudimetla @aravindhp we can use validation markers to handle this. Check [this](https://book.kubebuilder.io/reference/generating-crd.html). I tried it works.

File: test/e2e/wmco_test.go [hunk 12]
```diff
@@ -23,6 +25,32 @@ var (
 	timeout              = time.Minute * 60
 	cleanupRetryInterval = time.Second * 1
 	cleanupTimeout       = time.Second * 5
+
+	wmcReplicasFieldValidationTests = []struct {
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: fixed

File: test/e2e/wmco_test.go [hunk 13]
```diff
@@ -131,3 +160,60 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+						Namespace: namespace,
+					},
+					Spec: operator.WindowsMachineConfigSpec{
+						InstanceType: ec2InstanceType,
+						AWS:          &operator.AWS{CredentialAccountID: awsAccountID, SSHKeyPair: sshKey},
+					},
+				}
+			}
+
+			err = framework.Global.Client.Create(context.TODO(), wmc,
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: Done

File: test/e2e/wmco_test.go (hunk 7 above)
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: Lets handle this when we rebase

File: test/e2e/wmco_test.go [hunk 14]
```diff
@@ -145,66 +161,59 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      wmcObjectMetaName,
+						Namespace: namespace,
+					},
+					Spec: operator.WindowsMachineConfigSpec{
+						InstanceType: ec2InstanceType,
+						AWS:          &operator.AWS{CredentialAccountID: awsAccountID, SSHKeyPair: sshKey},
+						Replicas:     test.replicasFieldValue,
+					},
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: Done

File: test/e2e/wmco_test.go [hunk 15]
```diff
@@ -145,66 +161,59 @@ func testWindowsNodeDeletion(t *testing.T) {
...
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "instance",
-			Namespace: namespace,
-		},
-		Spec: operator.WindowsMachineConfigSpec{
-			InstanceType: ec2InstanceType,
-			AWS:          &operator.AWS{CredentialAccountID: awsAccountID, SSHKeyPair: sshKey},
-		},
-	}
+	for _, test := range wmcReplicasFieldValidationTests {
```
[the remark this answers is not in the document]
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: My input is replicas field and output is error/no error, both of them are part of the table. Not sure what you mean by input here.

[the remark this answers is not in the document]
Comment (aravindhp) [reply test/e2e/wmco_test.go]: You are crafting the CR based on what test is running. Your input can be the CR itself rather a field in the CR.

File: test/e2e/wmco_test.go [hunk 16]
```diff
@@ -131,3 +146,68 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+func testWMCValidation(t *testing.T) {
+	var wmcReplicasFieldValidationTests = []struct {
+		name                       string
+		isReplicasFieldPresent     bool
+		replicasFieldValue         int
+		isTestExpectedToThrowError bool
+		expectedErrorInTest        string
+	}{
+		// We are only checking negative test cases here, positive test cases would check if custom resource is
+		// getting created as expected and they are handled elsewhere
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: Where exactly are they handled? I think it'd be better if you mention the test name or have this comment at the testWMCValidation function level

File: test/e2e/wmco_test.go [hunk 17]
```diff
@@ -131,3 +146,68 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+		{"replicas field absent", false, 0, false, ""},
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: Be explicit about the fields that you want to populate. Please look at example here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/helper/node_affinity_test.go#L34=#L38

File: test/e2e/wmco_test.go [hunk 18]
```diff
@@ -131,3 +146,68 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+		{"replicas field value less than 0", true, -1, true, "spec.replicas in body should be greater than or equal to 0"},
+	}
+
+	testCtx := framework.NewTestCtx(t)
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: This block can be moved to starting as there is not point in declaring the `wmcReplicasFieldValidationTests` struct without getting the namespace

File: test/e2e/wmco_test.go [hunk 19]
```diff
@@ -131,3 +146,68 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+				assert.Equal(t, test.replicasFieldValue, wmc.Spec.Replicas, "Replicas value of the  WMC custom resource "+
+					"is not as expected")
+			}
+		})
+	}
+	defer testCtx.Cleanup()
+}
+
+// createWindowsMachineConfig creates a WindowsMachineConfig object
+func createWindowsMachineConfig(namespace string, isReplicasFieldRequired bool, replicasFieldValue int) *operator.WindowsMachineConfig {
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: nit: This can be a method of `wmcReplicasFieldValidationTests` instead of passing around the arguments

File: test/e2e/wmco_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ravisantoshgudimetla) [reply test/e2e/wmco_test.go]: I don't have a strong opinion on this. Either way works as long as it is readable.

[the remark this answers is not in the document]
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: Went ahead with the approach @aravindhp suggested

File: test/e2e/wmco_test.go (hunk 16 above)
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: Done

File: test/e2e/wmco_test.go (hunk 17 above)
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: Done

File: test/e2e/wmco_test.go (hunk 18 above)
Comment (akhil-rane) [reply test/e2e/wmco_test.go]: Done

File: test/e2e/wmco_test.go [hunk 20]
```diff
@@ -131,3 +146,77 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+				&framework.CleanupOptions{TestContext: testCtx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+
+			if test.isTestExpectedToThrowError {
+				require.Error(t, err, "Creation of WMC custom resource did not throw an error when it was expected to")
+				assert.Contains(t, err.Error(), test.expectedErrorInTest,
+					"Creation of WMC custom resource threw an unexpected error")
+			} else {
+				require.NoError(t, err, "Creation of the WMC custom resource threw an error when it was expected not to")
+				// Fetching wmc persisted in etcd and checking if replicas field value is initialized as expected
+				persistedWmc := &operator.WindowsMachineConfig{}
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: `actualWMC ` instead of `persistedWMC`

File: test/e2e/wmco_test.go [hunk 21]
```diff
@@ -131,3 +146,77 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+		isTestExpectedToThrowError bool
+		expectedErrorInTest        string
+	}{
+		{
+			name:                       "replicas field absent",
+			wmc:                        createWindowsMachineConfig(namespace, false, 0),
+			isTestExpectedToThrowError: false,
+			expectedErrorInTest:        "",
+		},
+		{name: "replicas field value less than 0",
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go]: alignment.

File: test/e2e/wmco_test.go [hunk 22]
```diff
@@ -131,3 +146,77 @@ func testWindowsNodeDeletion(t *testing.T) {
...
+			err = framework.Global.Client.Create(context.TODO(), test.wmc,
+				&framework.CleanupOptions{TestContext: testCtx, Timeout: cleanupTimeout, RetryInterval: cleanupRetryInterval})
+
+			if test.isTestExpectedToThrowError {
+				require.Error(t, err, "Creation of WMC custom resource did not throw an error when it was expected to")
+				assert.Contains(t, err.Error(), test.expectedErrorInTest,
+					"Creation of WMC custom resource threw an unexpected error")
+			} else {
+				require.NoError(t, err, "Creation of the WMC custom resource threw an error when it was expected not to")
+				// Fetching wmc persisted in etcd and checking if replicas field value is initialized as expected
```
Comment (aravindhp) [inline test/e2e/wmco_test.go]: Lets be consistent when adding acronyms. `WMC`

File: test/e2e/validation_test.go [hunk 23]
```diff
@@ -196,3 +199,78 @@ func testValidateSecrets(t *testing.T, nodeCount int) {
...
+				actualWMC := &operator.WindowsMachineConfig{}
+				err = framework.Global.Client.Get(context.TODO(),
+					kubeTypes.NamespacedName{Name: wmcCRName, Namespace: namespace}, actualWMC)
+				require.NoError(t, err, "Could not get the WMC custom resource")
+				assert.Equal(t, test.wmc.Spec.Replicas, actualWMC.Spec.Replicas, "Replicas value of the  WMC custom "+
+					"resource is not as expected")
+			}
+		})
+	}
+	defer testCtx.Cleanup()
```
Comment (aravindhp) [inline test/e2e/validation_test.go]: What is the point in adding `defer` when it is the last call made in the method? You should do it right after the `testCtx` is initialized.

Comment (akhil-rane) [reply test/e2e/validation_test.go]: Done

File: pkg/apis/wmc/v1alpha1/windowsmachineconfig_types.go [hunk 24]
```diff
@@ -11,6 +11,8 @@ import (
 type WindowsMachineConfigSpec struct {
 	// Replicas represent how many Windows nodes to be added to the
 	// OpenShift cluster
+	// +kubebuilder:validation:Required
```
Comment (sebsoto) [inline pkg/apis/wmc/v1alpha1/windowsmachineconfig_types.go]: This didnt cause a change in the CRD. is that expected?

Comment (akhil-rane) [reply pkg/apis/wmc/v1alpha1/windowsmachineconfig_types.go]: Good catch. Removed it. Thanks

Comment (ravisantoshgudimetla) [reply pkg/apis/wmc/v1alpha1/windowsmachineconfig_types.go]: This should've  Minimum:0 as well

File: pkg/apis/wmc/v1alpha1/windowsmachineconfig_types.go [hunk 25]
```diff
@@ -11,6 +11,7 @@ import (
 type WindowsMachineConfigSpec struct {
 	// Replicas represent how many Windows nodes to be added to the
 	// OpenShift cluster
+	// +kubebuilder:validation:Minimum=0
```
Comment (ravisantoshgudimetla) [inline pkg/apis/wmc/v1alpha1/windowsmachineconfig_types.go]: Ok this looks correct to me. I hit the submit review too quickly.

**Comment by PratikMahajan** [comment]
LGTM
